Handling of activations in different dtypes

davidarps opened the pull request:
The current NeuroX version handles activations with the dtype `float32`/full precision floats. I found it beneficial to work with activations in `float16` in all steps. The numerical difference between the two dtypes can be neglected: For a small example, the mean difference between two activation vectors in both dtypes is 6*10^{-7}. 
The advantage is that, in float16, activations stored in hdf5 take half as much disk space as float32 and are written/loaded faster. Also, activation vectors in float16 take up only half as much RAM. So with RAM restrictiions, you can experiment with twice as much data.

The default dtype is still float32. So if no dtype is specified explicitly, behavior does not change.

Changes

- `transformers_extractor.py` specify dtype when getting activations from model
- `writer.py` specify dtype for writing activations to file. The dtype is a parameter for the abstract creation of an ActivationWriter and passed to the `JSONActivationsWriter.write_activations()`, even though the dtype does not matter when writing to json. Is this the best choice?
- `loader.py` Only changed the loading of hdf5 and json. Are the other formats in use as well? There are no writers for them.
- `utils.py.create_tensors` 
- `linear_probe` Training and evaluation. 
    - Training on CPU in float16 is afaik not recommended. Also on GPU, the model input must be converted to float32 even in `torch.cuda.amp.autocast()`. To avoid having to convert the whole training activations `X_train` to float32, only individual batches are converted to float32. Apart from that, the probe is trained in mixed precision on cuda. 
    - Evaluation: Similar issue as training, see comment.
- tests for utils and linear probe extended.

davidarps committed:
optional dtypes

dtype can explicitly specified in all extraction, writing and loading, probe training and probe eval.
Also, probes are trained with mixed precision

fdalvi commented on neurox/data/extraction/transformers_extractor.py:
```diff
@@ -103,10 +103,12 @@ def aggregate_repr(state, start, end, aggregation):
-            "This probably implies a special unicode character or text " +
-            "encoding issue in your original data that was dropped by the " +
-            "transformer model's tokenizer.\n")
+        sys.stderr.write(
```
As mentioned in #15, I'll run the formatting after these PRs are merged, so feel free to revert these lines back so that only relevant code is part of the PR. You don't have to strictly do this however!

fdalvi commented on neurox/data/extraction/transformers_extractor.py:
```diff
@@ -173,7 +176,7 @@ def extract_sentence_representations(
                 for hidden_states in all_hidden_states[1:]
             ]
-        all_hidden_states = np.array(all_hidden_states)
+        all_hidden_states = np.array(all_hidden_states).astype(dtype)
```
Slightly more efficient to specify dtype as an argument to np.array (so `np.array(all_hidden_states, dtype=dtype)`

davidarps replied:
Done

fdalvi commented on neurox/data/writer.py:
```diff
@@ -166,11 +178,17 @@ def write_activations(self, sentence_idx, extracted_words, activations):
+                    sentence_idx,
+                    extracted_words,
+                    activations[layer_idx, :, :],
+                    dtype=self.dtype,
```
Can you add `dtype` as a member of the base `ActivationsWriter` class? this way the `write_activations` signature will not need to be changed (`write_activations` can use `self.dtype` instead of using an argument)

Conceptually, arguments of `write_activations` must be things that may change with every write, and not things that are constant for the current session like dtype

davidarps replied:
Done. dtype is still be passed around during creation and opening between ActivationsWriter, ActivationsWiterManager and the HDF5ActivationsWriter, but the dtype needs to be passed by the user only once during creation

fdalvi commented on neurox/interpretation/linear_probe.py:
```diff
@@ -89,6 +94,9 @@ def _train_probe(
     is trained with Cross Entropy loss for classification tasks and a linear
     regression model is trained with MSE loss for regression tasks. The
     optimizer used is Adam with default ``torch.optim`` hyperparameters.
+    On GPU, the probe is trained with mixed precision (amp).
```
As i understand from your comment in the PR and here, GPU training requires float32, and CPU training as well. If the model is float32, and the batch is float32, why do we need mixed-precision training here?

fdalvi replied:
This is important, as the `scaler` will be used regardless if someone wants float16 or not, which changes the default behavior. I assume scaler with float32 won't change anything, but we need to be confident about this. Your insight and expertise is appreciated here!

davidarps replied:
I added mixed precision training because I assumed it would slightly accelerate training. Thinking about it again, it is not strictly needed, and indeed the default behavior might change. I removed it again. That means that now, probe training always uses full precision by converting individual batches to float32 if necessary.

fdalvi commented on neurox/interpretation/utils.py:
```diff
@@ -139,7 +143,14 @@ def count_target_words(tokens):
+    task_type="classification",
+    binarized_tag=None,
+    balance_data=False,
+    x_dtype=None,
```
I think it is clear that the dtype here refers to `X` (since labels would be an np.int array regardless). If you still think the distinction is important, please capitalize X (`X_dtype`)

davidarps replied:
changed to `dtype`, method description specifies that it belongs to X only

fdalvi commented on tests/interpretation/test_linear_probe.py:
```diff
@@ -10,27 +10,26 @@
 
 import neurox.interpretation.linear_probe as linear_probe
 
+
```
Unfortunately because of bad instructions, you ended up formatting this very poorly formatted file; Sorry for that! Do you mind pushing your changes pre-formatting so the diff is easier to review?

fdalvi replied:
I just noticed that the formatting is in a nice clean commit, so it should be as easy as `git revert <formatting-commit-id>` and pushing!

fdalvi reviewed (commented):
Again, thanks for your effort and excellent work! I've left a few comments; let me know what you think!

Like the control task PR, if you can rebase on the current master that would be much appreciated!

fdalvi commented:
Also just quoting #15 incase you miss it there:

> For now, I have not formatted all of the repo code, because it will make updating your PRs a bit painful, so I'll do that once for the whole repo once we have your PRs in. So for the PRs themselves, just make sure the documentation looks alright and don't worry about running black or any formatting.

davidarps committed:
Merge remote-tracking branch 'origin/master' into precision

davidarps committed:
fix default x_dtype in create tensor

- broken test, probably broken during merge

davidarps committed:
more efficient dtype assignment during extraction

davidarps committed:
rm dtype from write_activations in writer

davidarps committed:
rename x_dtype to dtype

davidarps committed:
adapt linear_probe

- rm mixed precision
- convert probe to float() in evaluate_probe if necessary. No inplace operation, creates copy of the object

davidarps committed:
Merge remote-tracking branch 'origin/master' into precision

davidarps commented:
Thank you again Fahim for your detailed comments, and for merging the other PR! I think I addressed all of your comments. Also, I
- removed the formatting commit with a hard reset. It is probably easiest if you format everything in one go later
- merged the most recent master (including the control task), so everything is up to date. There are no conflicts between my two PRs
- slightly changed the evaluation method along with the probe training method. I hope it's easier to read now. 

In case there are more changes needed, please let me know!

fdalvi commented on neurox/data/writer.py:
```diff
@@ -151,11 +153,17 @@ def open(self, num_layers):
             for layer_idx in self.layers:
                 local_filename = f"{self.filename[:-5]}-layer{layer_idx}.{self.filename[-4:]}"
-                _writer = self.base_writer(local_filename)
+                if self.base_writer == HDF5ActivationsWriter:
```
I just took a quick look at how many decimal places can be accurately represented using a float16, and it seems like  anything smaller than 0.0005 cannot be precisely represented above a value of 1; so maybe dtype can play a role in the JSONWriter where we reduce the number of decimal places we save? I think currently currently we save 8 decimal places. What do you think?

Regardless, it might still be a good idea to pass in dtype and avoid this conditional here (more chances of issue if a new format is introduced in the future, or is someone or us decides to do something with dtype in the JSONWriter)

fdalvi replied:
This is what I found with a cursory search: https://stackoverflow.com/questions/872544/what-range-of-numbers-can-be-represented-in-a-16-32-and-64-bit-ieee-754-syste

If you can search a bit more and get a confident number that would be great (and maybe add it as a reference in the code itself)

davidarps replied:
This is a nontrivial (but interesting!) issue. [This table](https://en.wikipedia.org/wiki/Half-precision_floating-point_format#Precision_limitations) shows the error intervals that depend on the absolute value of the half-precision float. With representations for a few sentences in BERT, I experienced a mean error of 6*10^{-7} between float16 and float32, with similarly low standard deviation of the errors. So that's quite precise. However, as you might know, some LMs have outlier neurons with abs(value)>10 (usually high negative activations). For these neurons, precision according to the table is $2^{-7} \approx 8*10^{-3}$ - much less precise than for values around 1. So the decimal places you save could really depend on the LM, your use case and the particular value of a neuron.

Possible solution: (i) Pass the dtype to the JSONActivationsWriter to avoid the if-statement, and (ii) save only 6 decimal places when dtype is float16? The latter change would roughly correspond to the mean error experienced for BERT, but it is a bit intransparent. So it could also be left out.

Should I put the link in one of the method docstrings of the transformers_extractor or writer? Or is a section in the docs needed that discusses which dtype to use? Some users might not think about the issue, and might want some guidance. Not sure where to put that though.

fdalvi replied:
I looked into the link you sent and also wrote a little script to see the differences (see end of message); as you say, we need approx 7 decimal digits to retain precision < 1 in float16. Doing some quick back of the envelope calculations:

For a bert model we have ~10k neurons from all layers
Lets say we have about 1000 sentences with 5 tokens each, so total 5k tokens

Number of activation values:
`5000 * 10000 = 50,000,000`

Using one less character for each of them (not strictly true because we can save < 8 decimal even now if the activation value does not have 8 decimal places, but for the sake of argument):
`50,000,000 bytes = ~50MB`

So 50MB saved for about 1000 sentences, which I'm not 100% is worth the intransparency you suggest. What do you think?

We should definitely pass the dtype to JSONWriter anyways even if we don't do anything with it.

```python
import numpy as np

np.set_printoptions(suppress=True)

# Check precisions
# Row 0: 0.xxxx
# Row 1: 1.xxxx
# Row 2: 10.xxxx
# Row 3: 100.xxxx
full_precision = np.zeros((4, 20), dtype="float32")
half_precision = np.zeros((4, 20), dtype="float16")

full_precision[1, :] = 1
half_precision[1, :] = 1

full_precision[2, :] = 10
half_precision[2, :] = 10

full_precision[3, :] = 100
half_precision[3, :] = 100

for i, e in enumerate(range(0, -20, -1)):
	full_precision[:, i] += 10 ** e
	half_precision[:, i] += 10 ** e

print(full_precision)
print(half_precision)
```

davidarps replied:
I always work with hdf5, so I can't speak about the json usecase in practics. But I agree to your calculation: 50MB in your example is only a tiny fraction of the total file size for storing activations for 1000 sentences in json. So I agree, this small difference is probably not worth it. I'll pass the dtype parameter to the json writer but do nothing with it.

fdalvi commented on neurox/interpretation/linear_probe.py:
```diff
@@ -168,21 +171,24 @@ def _train_probe(
-                + lambda_l2 * l2_penalty(weights)
-            )
+            
+            with torch.cuda.amp.autocast(): # for mixed precision
```
Do we need the autocast() now given that we are not doing mixed precision training?

davidarps replied:
No, removed. Sorry, I overlooked that.

fdalvi commented on neurox/interpretation/linear_probe.py:
```diff
@@ -89,6 +89,9 @@ def _train_probe(
     is trained with Cross Entropy loss for classification tasks and a linear
     regression model is trained with MSE loss for regression tasks. The
     optimizer used is Adam with default ``torch.optim`` hyperparameters.
+    On GPU, the probe is trained with mixed precision (amp). 
```
I think this comment needs to be updated with our new non-mixed precision training approach?

davidarps replied:
Yes, adapted

fdalvi commented on neurox/interpretation/linear_probe.py:
```diff
@@ -381,6 +392,8 @@ class and their associated scores are also part of the dictionary.
 
     if use_gpu:
         probe = probe.cuda()
+        if X.dtype == 'float16': # cuda, fp16 in X: perform eval in half precision
```
So it seems like we can do evaluation in half precision but not training; If you think this gives us a reasonable speed up, we can leave this in; If not, we can just do full precision eval just like in train; I'll defer to your expertise here

davidarps replied:
The problem is in the current version of torch: On CPU, you should not use float16 for various reasons (see e.g. [this thread](https://github.com/pytorch/pytorch/issues/48245)). This is why CPU is treated differently than GPU here. I clarified the method comment.

fdalvi replied:
actually my question was at a higher level; given that we already do full precision training, is the benefit we get from "half precision evaluation on a GPU only" worth the added code complexity?

If you think that it is worth it, we can leave it in!

fdalvi commented on tests/interpretation/test_linear_probe.py:
```diff
@@ -149,6 +163,19 @@ def test_evaluate_probe_with_class_labels(self):
         self.assertIn("class0", scores)
         self.assertIn("class1", scores)
 
+    def test_evaluate_probe_with_class_labels_float16(self):
```
Can you make sure that these new tests cover all the new code that was added? I think they do but it would be great if you can confirm; You can run

```bash
python -m pytest -vv --cov --cov-report term-missing
```

to see the exact lines that are not covered by the test. I'll make this the default behavior for the future so that `run_tests` shows this information

davidarps replied:
I will check for that. BTW, you should make sure to add `pytest-cov` to the dependencies of the environment if you make this default behavior.

davidarps replied:
I found that only the part is not tested where in linear_probe.evaluate_probe() the execution depends on GPU availability. But GPU availability was also not tested before. Not sure it can be tested easily?

fdalvi replied:
pytest-cov is already part of the dev dependencies (https://github.com/fdalvi/NeuroX/blob/9829c8208d3181476ef9d60be725a37d446d2694/setup.cfg#L41), maybe you had created the env when it was not there yet.

Yes, you can ignore the missing GPU tests, they are indeed a bit tricky to mock for testing!

davidarps replied:
Regarding evaluation: 
I played a bit with `python -m timeit` and dtype conversion on an average CPU, and a GPU. I found that the conversion from float16 to float32 does not take much time, maybe 100-200ms for converting BERT base activations for 1000 ten-word-sentences. This is barely noticeable in practice, so it's a good argument against the special case that is in the code now.

That being said, you need some additional code now because you need to make sure that X and the probe have the same dtype, and that the dtype is not float16 if you're on CPU. I suggest doing that by always enforcing evaluation in float32, regardless of the input formats. I pushed a simpler version of the eval method in commit 4cdc823d74aba55096155246435b53ac92e47a1d . Do you like it better as well?

fdalvi reviewed (commented):
Thanks David, most of the changes look good. I did have a few more comments and I've left them in the review.

fdalvi commented:
btw I had another random thought (we don't need to do anything about it, atleast not in this PR), but is the "default" these days float32 in numpy? What about torch? I had a feeling that it may be float64 already but I may be wrong.

davidarps committed:
give dtype to JSON writer even if it is not needed

davidarps committed:
no autocast and mixed prec. in training

davidarps commented:
Good point! The default in torch is float32, in numpy it's float64. That's why you have to be quite careful when creating tensors manually. In loader.py:99 (`load_activations`), it is checked that when loading hdf5 files and returning np arrays, the dtype is set to float32 if none is specified. So with this version, you only use float64 in the np arrays if you explicitly declare it.

davidarps committed:
clarify method comment

davidarps committed:
Merge remote-tracking branch 'origin/master' into precision

fdalvi commented on neurox/data/writer.py:
```diff
@@ -151,11 +153,14 @@ def open(self, num_layers):
+                if self.base_writer == HDF5ActivationsWriter:
+                    _writer = self.base_writer(local_filename, dtype=self.dtype)
+                else:
+                    _writer = self.base_writer(local_filename)
```
Given that JSONWriter now takes dtype as well, we don't need this if anymore right?

fdalvi commented on neurox/interpretation/linear_probe.py:
```diff
@@ -88,7 +88,9 @@ def _train_probe(
     regression model is trained with MSE loss for regression tasks. The
-    optimizer used is Adam with default ``torch.optim`` hyperparameters.
+    optimizer used is Adam with default ``torch.optim`` hyperparameters. 
+    The individual batches generated from the X_train inputs are converted to flaot32, such that
```
typo "flaot32", break the line to roughly align with the text while you are fixing this

fdalvi reviewed (commented):
I think one final push and we are good to go; there have been some more changes recently on the repo, so please merge them as well if not done already!

davidarps committed:
typo/format

davidarps committed:
rm special case for different writers

davidarps committed:
always evaluate in float32 regardless

fdalvi commented on neurox/data/writer.py:
```diff
@@ -218,13 +219,13 @@ def close(self):
                 f"Output filename ({self.filename}) does not end with .json, but output file type is json."
             )
-
+        self.dtype = dtype
```
This should be passed to the parent constructor (like `HDF5ActivationsWriter` `super()` call)

fdalvi replied:
I'll take care of this in a future commit

fdalvi reviewed (approved):
Thanks a lot David, I think this looks good and I'll merge it in!
